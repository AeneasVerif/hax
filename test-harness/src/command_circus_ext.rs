use lazy_static::{__Deref, lazy_static};
use std::{ffi::OsStr, process::Command};

pub trait CommandCircusExt {
    fn circus<I: IntoIterator<Item = S>, S: AsRef<OsStr>>(args: I) -> Self;
}

impl CommandCircusExt for Command {
    fn circus<I: IntoIterator<Item = S>, S: AsRef<OsStr>>(args: I) -> Command {
        use assert_cmd::cargo::cargo_bin;
        use std::path::PathBuf;
        struct Paths {
            driver: PathBuf,
            engine: PathBuf,
            cargo_circus: PathBuf,
        }
        const CARGO_CIRCUS: &str = "cargo-circus";
        lazy_static! {
            static ref PATHS: Option<Paths> = {
                if let "yes" | "y" | "true" | "1" = std::env::var("CARGO_TESTS_ASSUME_BUILT").unwrap_or("".into()).to_lowercase().as_str() {
                    return None;
                }
                let root = std::env::current_dir().unwrap();
                let root = root.parent().unwrap();
                let engine_dir = root.join("engine");
                // Make sure binaries are built
                assert!(Command::new("cargo")
                        .args(&["build", "--workspace", "--bins"])
                        .status()
                        .unwrap()
                        .success());
                assert!(Command::new("dune")
                        .args(&["build"])
                        .env("CIRCUS_JSON_SCHEMA_EXPORTER_BINARY", cargo_bin("circus-export-json-schemas"))
                        .current_dir(engine_dir.clone())
                        .status()
                        .unwrap()
                        .success());
                Some(Paths {
                    driver: cargo_bin("driver-circus-frontend-exporter"),
                    cargo_circus: cargo_bin(CARGO_CIRCUS),
                    engine: engine_dir.join("_build/install/default/bin/circus-engine"),
                })
            };
        }
        let mut cmd = match PATHS.deref() {
            Some(paths) => {
                let mut cmd = Command::new(paths.cargo_circus.clone());
                cmd.env("CIRCUS_RUSTC_DRIVER_BINARY", paths.driver.clone());
                cmd.env("CIRCUS_ENGINE_BINARY", paths.engine.clone());
                cmd
            }
            None => Command::new(CARGO_CIRCUS),
        };
        cmd.args(args);
        // As documented in
        // https://doc.rust-lang.org/cargo/reference/environment-variables.html#dynamic-library-paths,
        // [cargo run] (and thus also [cargo test]) sets dynamic
        // library paths, which causes some issues with dependencies
        // when compiling without rustup
        for env in ["DYLD_FALLBACK_LIBRARY_PATH", "LD_LIBRARY_PATH"] {
            cmd.env_remove(env);
        }
        cmd
    }
}
