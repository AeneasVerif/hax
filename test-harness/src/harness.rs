mod command_circus_ext;
use command_circus_ext::*;
use serde_json::Value;
use std::process::{Command, Stdio};

#[derive(Clone, Debug)]
pub enum TestKind {
    Translate { backend: String },
    Lint { linter: String },
}

impl TestKind {
    fn as_subcommands(&self) -> [String; 2] {
        match self {
            TestKind::Lint { linter } => ["lint".to_string(), linter.clone()],
            TestKind::Translate { backend } => ["into".to_string(), backend.clone()],
        }
    }
}

#[derive(Clone, Debug, serde::Serialize)]
pub struct TestSpec {
    /// a slow test can be made optional
    pub optional: bool,
    /// a broken test a test that should succeed (or fail) but does
    /// not dues to a bug to be fixed (see field [issue_id] below)
    pub broken: bool,
    /// Github issue ID
    pub issue_id: Option<u64>,
    /// Is that a positive or a negative test?
    pub positive: bool,
}
impl From<Value> for TestSpec {
    /// Parse a JSON value into a TestSpec
    fn from(o: Value) -> Self {
        let as_bool = |key: &str, default: bool| {
            let b = &o[key];
            if b.is_null() {
                default
            } else {
                b.as_bool().unwrap()
            }
        };
        Self {
            optional: as_bool("optional", false),
            positive: as_bool("positive", true),
            broken: as_bool("broken", false),
            issue_id: o["issue_id"].as_u64(),
        }
    }
}

/// The information for a test is given by `cargo metadata`
#[derive(Clone, Debug)]
pub struct TestInfo {
    pub name: String,
    pub manifest: std::path::PathBuf,
    pub description: Option<String>,
}

#[derive(Clone, Debug)]
pub struct Test {
    pub kind: TestKind,
    pub info: TestInfo,
    pub spec: TestSpec,
}

impl std::fmt::Display for Test {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{} - {:?}", self.info.name, self.kind)?;
        if let Some(issue_id) = self.spec.issue_id {
            write!(f, " #{}", issue_id)?;
        };
        Ok(())
    }
}

impl Test {
    fn into_runner(self, workspace: String) -> Result<(), libtest_mimic::Failed> {
        // 1. cook a command
        let mut cmd = Command::circus(&["-C"]);
        cmd.arg("--manifest-path").arg(self.info.manifest.clone());
        cmd.arg(";");
        cmd.stdout(Stdio::piped()).stderr(Stdio::piped());
        cmd.args(self.kind.as_subcommands());

        // 2. execute it (twice, idea of @franziskuskiefer, so that
        // the messages related to building dependencies are not
        // included in the second one)
        let _ = cmd.output().unwrap();
        let out = cmd.output().unwrap();

        let command_successful = out.status.success();
        let cleanup = |s: String| {
            s.replace(r"\", "/")
                .replace(&workspace, "WORKSPACE_ROOT")
                .replace("Blocking waiting for file lock on build directory", "")
                .trim()
                .to_string()
        };
        let serr = cleanup(String::from_utf8_lossy(&out.stderr).to_string());
        let sout = cleanup(String::from_utf8_lossy(&out.stdout).to_string());

        // 3. make sure the test is successful
        let err = |s: &str| {
            Err(format!(
                "Command {s}.\nThe command was: {:?}{}",
                cmd,
                if command_successful {
                    "".to_string()
                } else {
                    format!("\nSTDOUT:\n{}\nSTDERR:\n{}", sout, serr)
                }
            ))
        };
        match (command_successful, (self.spec.positive, self.spec.broken)) {
            (false, (false, false) | (true, true)) => {
                #[derive(serde::Serialize)]
                struct Output {
                    serr: String,
                    sout: String,
                    exit: i32,
                }
                let exit = out.status.code().unwrap_or(std::i32::MAX);
                let out = &Output { serr, sout, exit };
                let name = format!(
                    "{}+{}",
                    self.info.name,
                    self.kind.as_subcommands().join("-")
                );
                insta::with_settings!({
                    info => &self.spec,
                }, { Ok(insta::assert_toml_snapshot!(name, out)) })
            }
            (false, (false, true)) => err("failed, but this is a negative test marked broken")?,
            (false, (true, false)) => err("failed")?,
            (true, (true, false) | (false, true)) => Ok(()),
            (true, (true, true)) => err("succeeded, but this is a positive test marked broken")?,
            (true, (false, false)) => err("succeeded, but this is a negative test")?,
        }
    }

    fn into_trial(&self, workspace: &String) -> libtest_mimic::Trial {
        libtest_mimic::Trial::test(format!("{}", &self), {
            let test = self.clone();
            let workspace = workspace.clone();
            move || test.clone().into_runner(workspace)
        })
        .with_kind(if self.spec.positive {
            "positive"
        } else {
            "negative"
        })
        .with_ignored_flag(self.spec.optional)
    }
}

/// Given [metadata] the table declared in a test's [Cargo.toml]
/// [workspace.circus-tests], this function returns a list of tests
fn parse_circus_tests_metadata(info: TestInfo, metadata: &Value) -> Vec<Test> {
    metadata
        .as_object()
        .expect("Expected value at key [circus-tests] to be a dictionary")
        .into_iter()
        .flat_map(|(a, o)| {
            o.as_object()
                .expect(format!("Expected value at key [{}] be a dictionary", a).as_str())
                .into_iter()
                .flat_map(|(key, o)| key.split("+").map(|k| (k.trim().to_string(), o.clone())))
                .map(|(b, o)| (a.clone(), b, o))
        })
        .map(|(a, b, o)| Test {
            spec: o.into(),
            info: info.clone(),
            kind: match a.as_str() {
                "into" => TestKind::Translate { backend: b },
                "lint" => TestKind::Lint { linter: b },
                _ => panic!("unexpected metadata [circus-tests.{}.{}]", a, b),
            },
        })
        .collect()
}

fn main() {
    let metadata = cargo_metadata::MetadataCommand::new()
        .manifest_path("../tests/Cargo.toml")
        .exec()
        .unwrap();
    let workspace_root: String = metadata.workspace_root.into();

    libtest_mimic::run(
        &libtest_mimic::Arguments::from_args(),
        metadata
            .packages
            .into_iter()
            .flat_map(|o| {
                parse_circus_tests_metadata(
                    TestInfo {
                        name: o.name,
                        description: o.description,
                        manifest: o.manifest_path.into(),
                    },
                    &o.metadata["circus-tests"],
                )
            })
            .map(|test| test.into_trial(&workspace_root))
            .collect(),
    )
    .exit();
}
