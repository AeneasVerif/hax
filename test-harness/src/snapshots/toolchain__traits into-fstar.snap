---
source: test-harness/src/harness.rs
expression: snapshot
info:
  kind:
    Translate:
      backend: fstar
  info:
    name: traits
    manifest: traits/Cargo.toml
    description: ~
  spec:
    optional: false
    broken: false
    issue_id: ~
    positive: true
    snapshot:
      stderr: true
      stdout: true
    include_flag: ~
    backend_options: ~
---
exit = 0
stderr = '''
Compiling traits v0.1.0 (WORKSPACE_ROOT/traits)
disabled backtrace
warning[HaxFront]: Hax frontend found a projected type with escaping bound vars. Please report https://github.com/hacspec/hax/issues/495
                   
                   Context:
                    - alias_ty: AliasTy {
                       substs: [
                           P,
                           (&u8,),
                       ],
                       def_id: DefId(2:2933 ~ core[7e13]::ops::function::FnOnce::Output),
                   }
                    - alias_kind: Projection
                    - trait_ref: <P as std::ops::FnOnce<(&u8,)>>
                    - trait_ref_and_substs: (
                       <P as std::ops::FnOnce<(&u8,)>>,
                       [],
                   )
                    - rebased_substs: [
                       P,
                       (&u8,),
                       (&u8,),
                   ]
                    - norm_rebased_substs: Ok(
                       <P as std::ops::FnOnce<(&u8,)>>,
                   )
                    - norm_substs: Ok(
                       <P as std::ops::FnOnce<(&u8,)>>,
                   )
                    - early_binder_substs: Ok(
                       <P as std::ops::FnOnce<(&u8,)>>,
                   )
                    - early_binder_rebased_substs: Ok(
                       <P as std::ops::FnOnce<(&u8,)>>,
                   )
   --> traits/src/lib.rs:107:13
    |
107 |             impl<P: FnMut(&u8) -> bool> Trait for P {}
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: ⚠️ This is a bug in Hax's frontend.
            Please report this error to https://github.com/hacspec/hax/issues with some context (e.g. the current crate)!

disabled backtrace
disabled backtrace
disabled backtrace
disabled backtrace
disabled backtrace
disabled backtrace
disabled backtrace
disabled backtrace
warning[HaxFront]: Hax frontend found a projected type with escaping bound vars. Please report https://github.com/hacspec/hax/issues/495
                   
                   Context:
                    - alias_ty: AliasTy {
                       substs: [
                           P,
                           (&<I as std::iter::Iterator>::Item,),
                       ],
                       def_id: DefId(2:2933 ~ core[7e13]::ops::function::FnOnce::Output),
                   }
                    - alias_kind: Projection
                    - trait_ref: <P as std::ops::FnOnce<(&<I as std::iter::Iterator>::Item,)>>
                    - trait_ref_and_substs: (
                       <P as std::ops::FnOnce<(&<I as std::iter::Iterator>::Item,)>>,
                       [],
                   )
                    - rebased_substs: [
                       P,
                       (&<I as std::iter::Iterator>::Item,),
                       (&<I as std::iter::Iterator>::Item,),
                   ]
                    - norm_rebased_substs: Err(
                       Type(
                           (&<P as std::iter::Iterator>::Item,),
                       ),
                   )
                    - norm_substs: Err(
                       Type(
                           (&<P as std::iter::Iterator>::Item,),
                       ),
                   )
                    - early_binder_substs: Ok(
                       <(&<I as std::iter::Iterator>::Item,) as std::ops::FnOnce<(&<P as std::iter::Iterator>::Item,)>>,
                   )
                    - early_binder_rebased_substs: Ok(
                       <(&<I as std::iter::Iterator>::Item,) as std::ops::FnOnce<(&<P as std::iter::Iterator>::Item,)>>,
                   )
  --> /nix/store/bbq0z3kg0b7hb3n7agk20r7hg3alf4kb-rust-default-1.72.0-nightly-2023-06-28/lib/rustlib/src/rust/library/core/src/iter/adapters/filter.rs:51:1
   |
51 | impl<I: Iterator, P> Iterator for Filter<I, P>
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: ⚠️ This is a bug in Hax's frontend.
           Please report this error to https://github.com/hacspec/hax/issues with some context (e.g. the current crate)!

disabled backtrace
disabled backtrace
disabled backtrace
warning: `traits` (lib) generated 2 warnings
    Finished dev [unoptimized + debuginfo] target(s) in XXs'''

[stdout]
diagnostics = []

[stdout.files]
"Traits.For_clauses.Issue_495_.Minimized_3_.fst" = '''
module Traits.For_clauses.Issue_495_.Minimized_3_
#set-options "--fuel 0 --ifuel 1 --z3rlimit 15"
open Core
open FStar.Mul

class t_Trait (v_Self: Type0) = { __marker_trait_t_Trait:Prims.unit }

[@@ FStar.Tactics.Typeclasses.tcinstance]
let impl
      (#v_P: Type0)
      (#[FStar.Tactics.Typeclasses.tcresolve ()] i0: Core.Ops.Function.t_FnMut v_P u8)
    : t_Trait v_P = { __marker_trait = () }
'''
"Traits.For_clauses.Issue_495_.fst" = '''
module Traits.For_clauses.Issue_495_
#set-options "--fuel 0 --ifuel 1 --z3rlimit 15"
open Core
open FStar.Mul

let minimized_1_ (list: Alloc.Vec.t_Vec u8 Alloc.Alloc.t_Global)
    : Alloc.Vec.t_Vec u8 Alloc.Alloc.t_Global =
  Core.Iter.Traits.Iterator.f_collect (Core.Iter.Traits.Iterator.f_filter ({
            Core.Ops.Range.f_start = 0uy;
            Core.Ops.Range.f_end = 5uy
          }
          <:
          Core.Ops.Range.t_Range u8)
        (fun temp_0_ ->
            let _:u8 = temp_0_ in
            true)
      <:
      Core.Iter.Adapters.Filter.t_Filter (Core.Ops.Range.t_Range u8) (u8 -> bool))

let minimized_2_ (it: Core.Iter.Adapters.Filter.t_Filter (Core.Ops.Range.t_Range u8) (u8 -> bool))
    : Prims.unit =
  let (v__indices: Alloc.Vec.t_Vec u8 Alloc.Alloc.t_Global):Alloc.Vec.t_Vec u8 Alloc.Alloc.t_Global
  =
    Core.Iter.Traits.Iterator.f_collect it
  in
  ()

let original_function_from_495_ (list: Alloc.Vec.t_Vec u8 Alloc.Alloc.t_Global) : Prims.unit =
  let (v__indices: Alloc.Vec.t_Vec u8 Alloc.Alloc.t_Global):Alloc.Vec.t_Vec u8 Alloc.Alloc.t_Global
  =
    Core.Iter.Traits.Iterator.f_collect (Core.Iter.Traits.Iterator.f_filter ({
              Core.Ops.Range.f_start = 0uy;
              Core.Ops.Range.f_end = 5uy
            }
            <:
            Core.Ops.Range.t_Range u8)
          (fun i ->
              let i:u8 = i in
              let _, out:(Core.Slice.Iter.t_Iter u8 & bool) =
                Core.Iter.Traits.Iterator.f_any (Core.Slice.impl__iter (Core.Ops.Deref.f_deref list
                        <:
                        t_Slice u8)
                    <:
                    Core.Slice.Iter.t_Iter u8)
                  (fun n ->
                      let n:u8 = n in
                      n =. i <: bool)
              in
              out)
        <:
        Core.Iter.Adapters.Filter.t_Filter (Core.Ops.Range.t_Range u8) (u8 -> bool))
  in
  ()
'''
"Traits.For_clauses.fst" = '''
module Traits.For_clauses
#set-options "--fuel 0 --ifuel 1 --z3rlimit 15"
open Core
open FStar.Mul

class t_Foo (v_Self: Type0) (v_T: Type0) = {
  f_to_t_pre:v_Self -> bool;
  f_to_t_post:v_Self -> v_T -> bool;
  f_to_t:x0: v_Self -> Prims.Pure v_T (f_to_t_pre x0) (fun result -> f_to_t_post x0 result)
}

let v__f (#v_X: Type0) (#[FStar.Tactics.Typeclasses.tcresolve ()] i1: t_Foo v_X u8) (x: v_X)
    : Prims.unit =
  let _:u8 = f_to_t x in
  ()
'''
"Traits.fst" = '''
module Traits
#set-options "--fuel 0 --ifuel 1 --z3rlimit 15"
open Core
open FStar.Mul

class t_Bar (v_Self: Type0) = {
  f_bar_pre:v_Self -> bool;
  f_bar_post:v_Self -> Prims.unit -> bool;
  f_bar:x0: v_Self -> Prims.Pure Prims.unit (f_bar_pre x0) (fun result -> f_bar_post x0 result)
}

let impl_2__method (#v_T: Type0) (#[FStar.Tactics.Typeclasses.tcresolve ()] i1: t_Bar v_T) (x: v_T)
    : Prims.unit = f_bar x

type t_Error = | Error_Fail : t_Error

let impl__Error__for_application_callback (_: Prims.unit) :  Prims.unit -> t_Error =
  fun temp_0_ ->
    let _:Prims.unit = temp_0_ in
    Error_Fail <: t_Error

let t_Error_cast_to_repr (x: t_Error) : isize = match x with | Error_Fail  -> isz 0

class t_Lang (v_Self: Type0) = {
  f_Var:Type0;
  f_s_pre:v_Self -> i32 -> bool;
  f_s_post:v_Self -> i32 -> (v_Self & f_Var) -> bool;
  f_s:x0: v_Self -> x1: i32
    -> Prims.Pure (v_Self & f_Var) (f_s_pre x0 x1) (fun result -> f_s_post x0 x1 result)
}

class t_SuperTrait (v_Self: Type0) = {
  [@@@ FStar.Tactics.Typeclasses.no_method]_super_10391689928755043351:Core.Clone.t_Clone v_Self;
  f_function_of_super_trait_pre:v_Self -> bool;
  f_function_of_super_trait_post:v_Self -> u32 -> bool;
  f_function_of_super_trait:x0: v_Self
    -> Prims.Pure u32
        (f_function_of_super_trait_pre x0)
        (fun result -> f_function_of_super_trait_post x0 result)
}

[@@ FStar.Tactics.Typeclasses.tcinstance]
let impl_SuperTrait_for_i32: t_SuperTrait i32 =
  {
    _super_10391689928755043351 = FStar.Tactics.Typeclasses.solve;
    f_function_of_super_trait_pre = (fun (self: i32) -> true);
    f_function_of_super_trait_post = (fun (self: i32) (out: u32) -> true);
    f_function_of_super_trait = fun (self: i32) -> cast (Core.Num.impl__i32__abs self <: i32) <: u32
  }

class t_Foo (v_Self: Type0) = {
  f_AssocType:Type0;
  f_AssocType_1566547207180513319:t_SuperTrait f_AssocType;
  f_AssocType_18427724250319916167:Core.Clone.t_Clone f_AssocType;
  f_N:usize;
  f_assoc_f_pre:Prims.unit -> bool;
  f_assoc_f_post:Prims.unit -> Prims.unit -> bool;
  f_assoc_f:x0: Prims.unit
    -> Prims.Pure Prims.unit (f_assoc_f_pre x0) (fun result -> f_assoc_f_post x0 result);
  f_method_f_pre:v_Self -> bool;
  f_method_f_post:v_Self -> Prims.unit -> bool;
  f_method_f:x0: v_Self
    -> Prims.Pure Prims.unit (f_method_f_pre x0) (fun result -> f_method_f_post x0 result);
  f_assoc_type_pre:{| i3: Core.Marker.t_Copy f_AssocType |} -> f_AssocType -> bool;
  f_assoc_type_post:{| i3: Core.Marker.t_Copy f_AssocType |} -> f_AssocType -> Prims.unit -> bool;
  f_assoc_type:{| i3: Core.Marker.t_Copy f_AssocType |} -> x0: f_AssocType
    -> Prims.Pure Prims.unit (f_assoc_type_pre i3 x0) (fun result -> f_assoc_type_post i3 x0 result)
}

let closure_impl_expr
      (#v_I: Type0)
      (#[FStar.Tactics.Typeclasses.tcresolve ()] i1: Core.Iter.Traits.Iterator.t_Iterator v_I)
      (it: v_I)
    : Alloc.Vec.t_Vec Prims.unit Alloc.Alloc.t_Global =
  Core.Iter.Traits.Iterator.f_collect (Core.Iter.Traits.Iterator.f_map it (fun x -> x)
      <:
      Core.Iter.Adapters.Map.t_Map v_I (Prims.unit -> Prims.unit))

let closure_impl_expr_fngen
      (#v_I #v_F: Type0)
      (#[FStar.Tactics.Typeclasses.tcresolve ()] i2: Core.Iter.Traits.Iterator.t_Iterator v_I)
      (#[FStar.Tactics.Typeclasses.tcresolve ()] i3: Core.Ops.Function.t_FnMut v_F Prims.unit)
      (it: v_I)
      (f: v_F)
    : Alloc.Vec.t_Vec Prims.unit Alloc.Alloc.t_Global =
  Core.Iter.Traits.Iterator.f_collect (Core.Iter.Traits.Iterator.f_map it f
      <:
      Core.Iter.Adapters.Map.t_Map v_I v_F)

let f (#v_T: Type0) (#[FStar.Tactics.Typeclasses.tcresolve ()] i1: t_Foo v_T) (x: v_T) : Prims.unit =
  let _:Prims.unit = f_assoc_f () in
  f_method_f x

let g (#v_T: Type0) (#[FStar.Tactics.Typeclasses.tcresolve ()] i1: t_Foo v_T) (x: i1.f_AssocType)
    : u32 = f_function_of_super_trait x

[@@ FStar.Tactics.Typeclasses.tcinstance]
let impl_Foo_for_tuple_: t_Foo Prims.unit =
  {
    f_AssocType = i32;
    f_AssocType_1566547207180513319 = FStar.Tactics.Typeclasses.solve;
    f_N = sz 32;
    f_assoc_f_pre = (fun (_: Prims.unit) -> true);
    f_assoc_f_post = (fun (_: Prims.unit) (out: Prims.unit) -> true);
    f_assoc_f = (fun (_: Prims.unit) -> () <: Prims.unit);
    f_method_f_pre = (fun (self: Prims.unit) -> true);
    f_method_f_post = (fun (self: Prims.unit) (out: Prims.unit) -> true);
    f_method_f = (fun (self: Prims.unit) -> f_assoc_f ());
    f_assoc_type_pre = (fun (_: i32) -> true);
    f_assoc_type_post = (fun (_: i32) (out: Prims.unit) -> true);
    f_assoc_type = fun (_: i32) -> ()
  }

type t_Struct = | Struct : t_Struct
'''
